medusa/storage/azure_storage.py

```diff
@@ -1,81 +1,336 @@
             )
-            driver.connection.user_id = credentials['storage_account']
+        return blobs
+
+    def _get_blob_hash(self, bp: BlobProperties) -> str:
+        md5_hash = bp.get('content_settings', {}).get('content_md5', bp.etag)
+        if md5_hash == bp.etag:
+            md5_hash_str = md5_hash.replace('"', '')
+        elif md5_hash is None:
+            md5_hash_str = 'no-md5-hash'
         else:
-            driver = AzureBlobsStorageDriver(
-                key=credentials['storage_account'],
-                secret=credentials['key']
+            md5_hash_str = base64.encodebytes(md5_hash).decode('UTF-8').strip()
+        return md5_hash_str
+
+    def get_object(self, bucket_name, object_key: str) -> t.Optional[AbstractBlob]:
+        # Doesn't actually read the contents, just lists the thing
+        try:
+            loop = self._get_or_create_event_loop()
+            o = loop.run_until_complete(self._get_object(object_key))
+            return o
+        except ObjectDoesNotExistError:
+            return None
+
+    async def _get_object(self, object_key: str) -> AbstractBlob:
+        blob = await self._stat_blob(object_key)
+        return blob
+
+    @retry(stop_max_attempt_number=MAX_UP_DOWN_LOAD_RETRIES, wait_fixed=5000)
+    def upload_blobs(self, srcs: t.List[t.Union[Path, str]], dest: str) -> t.List[ManifestObject]:
+        loop = self._get_or_create_event_loop()
+        manifest_objects = loop.run_until_complete(self._upload_blobs(srcs, dest))
+        return manifest_objects
+
+    async def _upload_blobs(self, srcs: t.List[t.Union[Path, str]], dest: str) -> t.List[ManifestObject]:
+        coros = [self._upload_blob(src, dest) for src in map(str, srcs)]
+        manifest_objects = []
+        n = int(self.config.concurrent_transfers)
+        for chunk in [coros[i:i + n] for i in range(0, len(coros), n)]:
+            manifest_objects += await asyncio.gather(*chunk)
+        return manifest_objects
+
+    async def _upload_blob(self, src: str, dest: str) -> ManifestObject:
+        src_chunks = src.split('/')
+        parent_name, file_name = src_chunks[-2], src_chunks[-1]
+
+        # check if objects resides in a sub-folder (e.g. secondary index). if it does, use the sub-folder in object path
+        object_key = (
+            "{}/{}/{}".format(dest, parent_name, file_name)
+            if parent_name.startswith(".")
+            else "{}/{}".format(dest, file_name)
+        )
+
+        file_size = os.stat(src).st_size
+        logging.debug(
+            '[Azure Storage] Uploading {} ({}) -> azure://{}/{}'.format(
+                src, _human_readable_size(file_size), self.config.bucket_name, object_key
             )
+        )
 
-        return driver
+        with open(src, "rb") as data:
+            blob_client = await self.azure_container_client.upload_blob(
+                name=object_key,
+                data=data,
+                overwrite=True,
+                max_concurrency=4,
```

Comment:
**Todo:** after doing some benchmarks, it looks like 16 gives us better performance. Could you set this value?

Reply:
Pushed a change.